Add AssetPartitionWipeRange class

smackesey opened the pull request:
## Summary & Motivation

Add an `AssetPartitionWipeRange` class to complement `AssetKeyPartitionKey`. This is used upstack in partition wiping PRs.

Like `AssetKeyPartitionKey`, this uses `None` for the `partition_range` if there is no range-- this is interpreted as "the whole asset".

This is called `AssetPartitionWipeRange` instead of just `AssetPartitionRange` to scope this abstraction explicitly to partition wiping for the time being. Later I expect to rename this to just `AssetPartitionRange` or similar as part of a broader rename of partition abstractions.

## How I Tested These Changes

Existing test suite.

smackesey commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22245?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#22245** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22245?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#22244** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22244?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

schrockn commented:
Is this totally necessary for the stack? Don't want to rush this one.

smackesey commented:
>Is this totally necessary for the stack? Don't want to rush this one.

It's not 100% necessary (I could just use dicts or something) but it will make the code harder to follow without it and we surely need something like this in the codebase. It's not public and could easily be renamed later if we wanted to-- why not add here since there is a legit use case to support? Note that we're not storing anything like this object in the DB yet.

schrockn commented:
If this is just for the stack I'd rather have this more clearly named with a more speciailzed name for a single code path and then generalize the name later.

smackesey commented:
>If this is just for the stack I'd rather have this more clearly named with a more speciailzed name for a single code path and then generalize the name later.

Renamed to `AssetPartitionWipeRange` and updated the summary.

schrockn commented on python_modules/dagster/dagster/_core/definitions/events.py:
```diff
@@ -58,6 +62,19 @@ class AssetKeyPartitionKey(NamedTuple):
 
+# This is currently used only for the asset partition wipe codepath. In the future, we can rename
+# to AssetPartitionRange or similar for more general use.
+class AssetPartitionWipeRange(NamedTuple):
+    """An AssetKey with a partition range."""
+
+    asset_key: AssetKey
+    partition_range: PartitionKeyRange
+
+    @property
+    def is_universal(self) -> bool:
+        return self.partition_range == UNIVERSAL_PARTITION_KEY_RANGE
+
```
Also weary of this `UNIVERSAL_PARTITION_KEY_RANGE` without further examples of how it is actually used. 

If it is "universal" but I access `partition_range` directly, is that valid? I'd rather introduce as few as new things as possible here.

smackesey replied:
It was used upstack but I've removed it and now am just using `None` (as with `AssetKeyPartitionKey`) to represent the whole asset. We can make the change to "all assets are partitioned" in a coherent way later.
